File: Java/src/Letter Combinations of a Phone Number.java
public class Solution {
    public List<String> letterCombinations(String digits) {
        Map<Character, String> pad = new HashMap<>();
        pad.put('2',"abc");
        pad.put('3',"def");
        pad.put('4',"ghi");
        pad.put('5',"jkl");
        pad.put('6',"mno");
        pad.put('7',"pqrs");
        pad.put('8',"tuv");
        pad.put('9',"wxyz");
        
        if(digits.length()== 0) return new ArrayList(Arrays.asList(""));

        List<String> cb0 = letterCombinations(digits.substring(0,digits.length()-1));
        List<String> cb1 = new ArrayList<>();
        char c = digits.charAt(digits.length() - 1);
        String s = pad.get(c);

        for(int i=0; i< s.length(); i++){
            String sAppend = s.substring(i,i+1);
            for (String sPrevious: cb0 ){
                String newS = sPrevious + sAppend;
                cb1.add(newS);
            }
        }
        return cb1;
    }
}

File: Python/Letter Combinations of a Phone Number.py
class Solution:
    # @return a list of strings, [s1, s2]
    def letterCombinations(self, digits):
        # define a dictionary of the character <-> number projection
        pad = {'2':'abc', '3':'def', '4':'ghi', '5':'jkl', '6':'mno', '7':'pqrs', '8':'tuv', '9':'wxyz'}
        # if the input is empty
        if len(digits) == 0:
            return [""]

        comb0 = self.letterCombinations(digits[:-1])
        string1 = pad[digits[-1]]
        comb1 = []
        for s in string1:
            for c in comb0:
                cs = c+s
                comb1.append(cs)
        return comb1